Add an SPI to suppress marking and highlighting when scrolling to text fragments

TanmoyPanigrahi opened the pull request:
#### 53ee2c4ceb88891f76673f40795e4e291b865671
<pre>
Add an SPI to suppress marking and highlighting when scrolling to text fragments
<a href="https://bugs.webkit.org/show_bug.cgi?id=271196">https://bugs.webkit.org/show_bug.cgi?id=271196</a>
<a href="https://rdar.apple.com/124974910">rdar://124974910</a>

Reviewed by Megan Gardner.

Add functionality enabling users of the SPI to prevent the
persistent highlighting/marking of the fragment scrolled to

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollToFragment):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setupPageConfiguration:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration init]):
(-[WKWebViewConfiguration encodeWithCoder:]):
(-[WKWebViewConfiguration initWithCoder:]):
(-[WKWebViewConfiguration copyWithZone:]):
(-[WKWebViewConfiguration _scrollToTextFragmentIndicatorEnabled]):
(-[WKWebViewConfiguration _setScrollToTextFragmentIndicatorEnabled:]):
(-[WKWebViewConfiguration _scrollToTextFragmentMarkingEnabled]):
(-[WKWebViewConfiguration _setScrollToTextFragmentMarkingEnabled:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:

Canonical link: <a href="https://commits.webkit.org/277571@main">https://commits.webkit.org/277571@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5ced334ffcbb5958ad25b9ba45870ea4e5cd864f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47920 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27121 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50789 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50601 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43973 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50226 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32993 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24608 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38986 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48502 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24784 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41440 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20284 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22271 "Passed tests") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/42622 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 14.0; Skipping applying patch since patch_id isn't provided; Checked out pull request; Exiting early after 10 failures. 20 tests run. 2 failures; Uploaded test results; Running layout-tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5967 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/41216 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44272 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/43060 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52495 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/47411 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22962 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19314 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46295 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24232 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45340 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10583 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25025 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/54909 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23953 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11288 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
```diff
@@ -168,6 +168,10 @@ typedef NS_ENUM(NSUInteger, _WKContentSecurityPolicyModeForExtension) {
 
 @property (nonatomic, setter=_setMarkedTextInputEnabled:) BOOL _markedTextInputEnabled WK_API_AVAILABLE(macos(14.0), ios(17.0));
+
+@property (nonatomic, setter=_setScrollToTextFragmentIndicatorEnabled:) BOOL _scrollToTextFragmentIndicatorEnabled WK_API_AVAILABLE(macos(14.0), ios(17.0));
```
this should probably use WK_IOS_TBA and WK_MAC_TBA

pxlcoder commented on Tools/DumpRenderTree/TestOptions.cpp:
```diff
@@ -99,6 +99,7 @@ const TestFeatures& TestOptions::defaults()
             { "RequiresUserGestureForMediaPlayback", false },
             { "RequiresUserGestureForVideoPlayback", false },
             { "ScrollToTextFragmentIndicatorEnabled", false },
+            { "ScrollToTextFragmentMarkingEnabled", true },
```
Not necessary, since the default is `true`.

pxlcoder commented on Tools/WebKitTestRunner/TestOptions.cpp:
```diff
@@ -125,6 +125,7 @@ const TestFeatures& TestOptions::defaults()
             { "RequiresUserGestureForMediaPlayback", false },
             { "RequiresUserGestureForVideoPlayback", false },
             { "ScrollToTextFragmentIndicatorEnabled", false },
+            { "ScrollToTextFragmentMarkingEnabled", true },
```
Ditto.

See also, the comment near the top of this block:

```
// Please do not add new options here if they are not necessary (e.g.
// an experimental feature which gets enabled by default automatically)
// as it adds a small amount of unnecessary work per-test.
```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
```diff
@@ -458,6 +466,9 @@ - (id)copyWithZone:(NSZone *)zone
     configuration->_allowsInlinePredictions = self->_allowsInlinePredictions;
 
+    configuration->_scrollToTextFragmentMarkingEnabled = self->_scrollToTextFragmentMarkingEnabled;
+    configuration->_scrollToTextFragmentMarkingEnabled = self->_scrollToTextFragmentMarkingEnabled;
```
I think you meant to use `_scrollToTextFragmentIndicatorEnabled` on one of these lines 😅

megangardner commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
```diff
@@ -328,6 +332,8 @@ - (void)encodeWithCoder:(NSCoder *)coder
     [coder encodeInteger:self.userInterfaceDirectionPolicy forKey:@"userInterfaceDirectionPolicy"];
 #endif
+    [coder encodeBool:self._scrollToTextFragmentIndicatorEnabled forKey:@"scrollToTextFragmentIndicatorEnabled"];
+    [coder encodeBool:self._scrollToTextFragmentMarkingEnabled forKey:@"scrollToTextFragmentMarkingEnabled"];
```
Not sure if this is needed.

pxlcoder replied:
There's no reason to exclude it. We want encode/decode of `WKWebViewConfiguration` to produce equivalent objects.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
```diff
@@ -179,6 +179,8 @@ @implementation WKWebViewConfiguration {
     double _sampledPageTopColorMinHeight;
     BOOL _allowsInlinePredictions;
+    BOOL _scrollToTextFragmentIndicatorEnabled;
+    BOOL _scrollToTextFragmentMarkingEnabled;
```
There's an ongoing effort to move ivars from `WKWebViewConfiguration` to `API::PageConfiguration`, so we should do in this PR before merging.

See https://github.com/WebKit/WebKit/pull/26157 for examples.

Commit:
Add an SPI to suppress marking and highlighting when scrolling to text fragments
https://bugs.webkit.org/show_bug.cgi?id=271196
rdar://124974910

Reviewed by Megan Gardner.

Add functionality enabling users of the SPI to prevent the
persistent highlighting/marking of the fragment scrolled to

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollToFragment):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setupPageConfiguration:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration init]):
(-[WKWebViewConfiguration encodeWithCoder:]):
(-[WKWebViewConfiguration initWithCoder:]):
(-[WKWebViewConfiguration copyWithZone:]):
(-[WKWebViewConfiguration _scrollToTextFragmentIndicatorEnabled]):
(-[WKWebViewConfiguration _setScrollToTextFragmentIndicatorEnabled:]):
(-[WKWebViewConfiguration _scrollToTextFragmentMarkingEnabled]):
(-[WKWebViewConfiguration _setScrollToTextFragmentMarkingEnabled:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:

Canonical link: https://commits.webkit.org/277571@main
